src/common/decorators/one-at-time.ts

```diff
@@ -1,22 +1,100 @@
-export function OneAtTime<T extends (...args: any[]) => Promise<any>>() {
+import 'reflect-metadata';
+const stakingModuleId = Symbol('StakingModuleId');
+
+/**
+ * A decorator that marks a specific parameter in a method for identifying the staking module ID
+ */
+export function StakingModuleId(
+  target: any,
+  propertyKey: string | symbol,
+  parameterIndex: number,
+) {
+  const existingMetadata: number[] =
+    Reflect.getOwnMetadata(stakingModuleId, target, propertyKey) || [];
+
+  if (existingMetadata.length === 0) {
+    Reflect.defineMetadata(
+      stakingModuleId,
+      [parameterIndex],
+      target,
+      propertyKey,
+    );
+  } else {
+    throw new Error(
+      `StakingModuleId decorator can only be applied to one parameter in method ${String(
+        propertyKey,
+      )}. It is already applied to parameter index ${existingMetadata[0]}`,
+    );
+  }
+}
+
+/**
+ * A decorator factory that produces a method decorator ensuring a function executes one at a time.
+ * Calls to the decorated method are restricted so that only one instance can be executed concurrently,
+ * either globally or per staking module ID
+ * A stuck function with the OneAtTime decorator will prevent the next executions of this function.
+ * That is why a timeout is set. If the execution of the promise is stuck, a timeout will occur. The default timeout is 10 minutes.
+ */
+export function OneAtTime<T extends (...args: any[]) => Promise<any>>(
+  timeout = 600000,
+) {
```

Comment:
I agree that it would look better if we moved the timeout logic to a separate decorator, but it is safer to leave it in one decorator to prevent the use of OneAtTime without a timeout. The OneAtTime decorator should include timeout logic to prevent a promise from getting stuck. Reliability is more important than beauty